### PR TITLE
Feature/21660 effective dates

### PR DESCRIPTION
**Pull request by agger-magenta**
OK, så skulle den vist være der. Eller tæt på. Jeg vil gerne have input til flere tests, der skal rettes op.

### CODE REVIEW

File: interface_test/21660_test_put_update.sh [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+# Test configuration
+DIR=$(dirname ${BASH_SOURCE[0]})
+source $DIR/config.sh
+
+result=$(curl -k -H "Content-Type: application/json" -X POST -d "$(cat test_data/facet_opret.json)" $HOST_URL/klassifikation/facet)
+uuid=$(expr "$result" : '.*"uuid": "\([^"]*\)"')
+echo $uuid
+curl -i -k -H "Content-Type: application/json" -X PUT -d "$(cat test_data/facet_reduce_effective_time_21660.json)" $HOST_URL/klassifikation/facet/$uuid
```
[the remark this answers is not in the document]
Comment (agger-magenta) [reply interface_test/21660_test_put_update.sh]: I updated the test suite with some tests too, though I might be missing some. There's no active harm done in testing this from here as well, I believe.

File: oio_rest/oio_rest/oio_rest.py [hunk 2]
```diff
@@ -362,31 +361,62 @@ def put_object(cls, uuid):
...
+        input = cls.get_json()
+        if not input:
+            return jsonify({'uuid': None}), 400
+        # Get most common parameters if available.
+        note = typed_get(input, "note", "")
+        registration = cls.gather_registration(input)
+
+        # We have input, we have a registration & captured the note - now do
+        # the deed.
+        "Edit or passivate."
```
[the remark this answers is not in the document]
Comment (agger-magenta) [reply oio_rest/oio_rest/oio_rest.py]: That should be a comment, I'll change it.

File: doc/database_changes.rst [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+The tables, indexes, datatypes and functions in the database of
+Magenta's Mox/LoRa-actual state server are generated when the
+recreatedb.sh script is executed. NOTICE: You do not want to run
+recreatedb.sh on a database, that is already in production (i.e. that
+has production data in it ) - as it will be dropped . The script is
+intended to initialize a new database.
+
+The recreatedb.sh script runs a process, where configuration defined in
+settings.py for each OIO-type, combined with the listing of OIO-types a
```
[the remark this answers is not in the document]
Comment (agger-magenta) [reply doc/database_changes.rst]: No, this used to be in `settings.py` but has been moved to `db_structure.py` a long time ago. I'll update the text to reflect that. (The documentation of the DB change process was written by Thomas Kristensen in early 2016, and my contribution was limited to formatting according to our line length conventions and fixing a single error).

File: oio_rest/oio_rest/oio_rest.py [hunk 4]
```diff
@@ -362,31 +361,62 @@ def put_object(cls, uuid):
...
-                    cls.__name__, note, registration, uuid
+            # Edit.
+            request.api_operation = "Ret"
+            # TODO: This needs to be implemented!
+            db.create_or_import_object(cls.__name__, note, registration, uuid)
+
+            return jsonify({'uuid': uuid}), 200
+
+        # We should never get here.
+        assert(0 != 0)
```
[the remark this answers is not in the document]
Comment (agger-magenta) [reply oio_rest/oio_rest/oio_rest.py]: Indeed, all the branches of the `if` statement returns something. I'll remove it.

File: db/db-templating/templates/as_create_or_import.jinja.sql [hunk 5]
```diff
@@ -21,6 +21,9 @@ DECLARE
   {% endfor %}
   {{oio_type}}_relationer {{oio_type|title}}RelationType;
   auth_filtered_uuids uuid[];
+  does_exist boolean;
+  new_{{oio_type}}_registrering {{oio_type}}_registrering;
+  prev_{{oio_type}}_registrering {{oio_type}}_registrering;
```
Comment (cmoesgaard) [inline db/db-templating/templates/as_create_or_import.jinja.sql]: What is this variable used for?

Comment (agger-magenta) [reply db/db-templating/templates/as_create_or_import.jinja.sql]: Embarassingly, the `prev_*_registrering` is not used for anything. I believe that I thought I needed it to compare with the old one, but this turned out not to be necessary.

Comment (cmoesgaard) [reply db/db-templating/templates/as_create_or_import.jinja.sql]: I see. My SQL is a little rusty, so I was afraid it was something obvious that I had missed. :smile:

Comment (agger-magenta) [reply db/db-templating/templates/as_create_or_import.jinja.sql]: I'll remove it as part of the cleanup when all the review comments are in.

**Comment by agger-magenta** [comment]
Fine, I'll look at it next week! There is a quick test of the PUT mechanism in `interface_tests`.

There shouldn't be a real danger of previous virkningstider being inherited, since nothing is inherited from the old registration - but since that was one of the reasons for making these changes, it makes sense to test it.

File: doc/database_changes.rst [hunk 6]
```diff
@@ -0,0 +1,146 @@
+Working with database templates and patches for Magenta's Mox/LoRa-Actual State Server
+======================================================================================
+
+The tables, indexes, datatypes and functions in the database of
+Magenta's Mox/LoRa-actual state server are generated when the
+recreatedb.sh script is executed. NOTICE: You do not want to run
```
Comment (danchr) [inline doc/database_changes.rst]: I generally prefer to have file names, etc. as fixed-width text to make them stand out — but that's not a requirement, though…

Also, you might want to use a [warning admonition](http://docutils.sourceforge.net/docs/ref/rst/directives.html#specific-admonitions), or similar, to add emphasis to the `NOTICE`.

File: doc/database_changes.rst [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+recreatedb.sh on a database, that is already in production (i.e. that
+has production data in it ) - as it will be dropped . The script is
```
Comment (danchr) [inline doc/database_changes.rst]: Extraneous whitespace before the end parenthesis.

File: doc/database_changes.rst [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+intended to initialize a new database.
+
+The recreatedb.sh script runs a process, where configuration defined in
+db_structure.py for each OIO-type, combined with the listing of
+OIO-types a few places in recreatedb.sh, will trigger the generation of
+tables, indexes, datatypes and functions in the database, to support the
+handling of the OIO-types.
```
Comment (danchr) [inline doc/database_changes.rst]: You should probably break this sentence into separate parts — at the moment it's a bit convoluted and hard to parse.

File: doc/database_changes.rst [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+'templates' to generate the bulk of the code/structures in the database
+to support a given OIO-type. In a few cases (e.g. for OIO-type
+'Dokument'), the db_structure.py file contains some extra configuration
+that is utilized in very specific parts of a template, to accommodate
+for variations for that particular type. However, in a lot of cases, the
+handling of variations is done by applying patch-files to the output of
+the template process, generating a modified SQL-script, which is then
+executed on the DB-server. This has the advantage of not increasing the
+complexity of the jinja-templates to accommodate every other variation,
+which may only apply to one or few OIO-types. The downside is, that you
```
Comment (danchr) [inline doc/database_changes.rst]: No comma here.

File: doc/database_changes.rst [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+   definition is ready.
+
+5. If you are altering an existing patch, disable the application of
+   this particular patch temporarily in the recreatedb.sh script. Now,
+   re-run recreatedb.sh, leaving you with the 'raw' output of the
+   template process for the particular function in the 'generated-files'
+   folder.
+
+6. Create the actual .diff file by running the command below,
+   substituting function names as needed.
```
Comment (danchr) [inline doc/database_changes.rst]: You can use a code block for this: Ending the section with two colons and indenting the next one makes it automagically fixed width.

File: doc/database_changes.rst [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+
+    `diff --context=5  ./patches/as_create_or_import_tilstand.org.sql ./generated-files/as_create_or_import_tilstand.sql > patches/as_create_or_import_tilstand.sql.diff`
+
+7. Make sure that the patch is applied in the relevant section of
+   recreatedb.sh script. That is, undo what you did in step 5 or - if
+   you are targeting a previously unpatched function/file - introduce a
+   line similar to the one below (again, substituting names as needed)
```
Comment (danchr) [inline doc/database_changes.rst]: Likewise, perhaps replacing the parentheses with a comma.

File: doc/database_changes.rst [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+
+11. Commit the files to git: The .sql file in the 'generated-files'
+    folder, the .org file in the 'patches' folder and the '.diff' file
+    also in the patches folder, the modified recreatedb.sh script and
+    files containing any new or modified tests.
+
+How changes in db_structure.py or jinja-templates may affect patching
+---------------------------------------------------------------------
+
+When you are introducing changes in db_structure.py or in
```
Comment (danchr) [inline doc/database_changes.rst]: I'd suggest either “When you introduce…” or “When introducing…”.

File: doc/database_changes.rst [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+jinja-tempaltes, you need to pay special attention, when your changes
```
Comment (danchr) [inline doc/database_changes.rst]: Splelnig.

File: doc/database_changes.rst [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+involves OIO-types, the function definitions of which are the target of
+patching.
+
+As a basis, you should always run recreatedb.sh and verify that all
+patches applied successfully, by inspecting the output of the script -
+and that all tests are still passing, by running run_tests.sh.
+
+However, if you are unsure, if the changes introduced may affect code in
```
Comment (danchr) [inline doc/database_changes.rst]: Leave out the second comma.

File: doc/database_changes.rst [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+relation to the relevant section in db_structure.py and you have
+completed the step described above - you should search the patch (.diff)
+files related to 'Tilstand' for the name of an existing relation type of
+unlimited cardinality. If you find occurrences in the diff files (and at
+the time of this writing, you should) of the 'sister'-relation type, you
+need to inspect the code in the full context and determine if not your
+new relation type, should be named in that particular section as well.
+Most likely, is should.
+
+As always, it is sound practise, to run the changes you wish to make by
```
Comment (danchr) [inline doc/database_changes.rst]: Kindly euthanise the comma before “to”.

File: doc/database_changes.rst [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+the developer available with the most in-depth knowledge of the code, to
+help identify aspects to be aware of.
```
Comment (danchr) [inline doc/database_changes.rst]: This sentence doesn't quite read well — I'd suggest rephrasing it, and adjusting it to emphasise code review. Something like:

> Please note that code review is even more important when making changes to the database, and you should take particular care to discuss any changes with whomever has the most in-depth knowledge of the code.

File: oio_rest/oio_rest/db.py [hunk 17]
```diff
@@ -371,6 +371,8 @@ def create_or_import_object(class_name, note, registration,
     # Do not supply date, that is generated by the DB.
     life_cycle_code = (Livscyklus.OPSTAAET.value if uuid is None
                        else Livscyklus.IMPORTERET.value)
+    if uuid and object_exists(class_name, uuid):
+        life_cycle_code = Livscyklus.RETTET.value
```
Comment (danchr) [inline oio_rest/oio_rest/db.py]: In my opinion, combining this line with the two previous ones into a single, three-case `if` would make the code more readable.

File: oio_rest/oio_rest/db.py [hunk 18]
```diff
@@ -442,18 +449,10 @@ def delete_object(class_name, registration, note, uuid):
     try:
         cursor.execute(sql)
     except psycopg2.Error as e:
-        not_found_msg = (
-            'Unable to update {} with uuid [{}], '
-            'being unable to find any previous registrations.\n'
-        ).format(class_name.lower(), uuid)
         already_deleted_msg = (
```
Comment (danchr) [inline oio_rest/oio_rest/db.py]: Why keep the check for `already_deleted_msg`? Are we still testing it?

File: oio_rest/oio_rest/oio_rest.py [hunk 19]
```diff
@@ -327,7 +327,7 @@ def gather_registration(cls, input):
     @requires_auth
     def put_object(cls, uuid):
         """
-        UPDATE, IMPORT or PASSIVIZE an  object.
+        IMPORT or UPDATE an  object, replacing its contents completely.
```
Comment (danchr) [inline oio_rest/oio_rest/oio_rest.py]: Perhaps `s/completely/entirely/`?

File: oio_rest/tests/test_integration_class.py [hunk 20]
```diff
@@ -132,7 +132,7 @@ def test_edit(self):
                 'uuid': objid,
             },
             json=util.get_fixture('klasse_opdater.json'),
-            method='PUT',
+            method='PATCH',
```
Comment (danchr) [inline oio_rest/tests/test_integration_class.py]: I can't help but worry that you've changed two of our integration tests to use `PATCH` without adding cases for `PUT`. I'd like to see integration tests for both behaviours — I believe `aktivitet` has a few unused fixtures copied over from the `interface_tests` that might be useful.

File: doc/index.rst [hunk 21]
```diff
@@ -16,6 +16,7 @@ Welcome to the LoRa MOX documentation!
    API-NOTES.rst
    praesentation.rst
    log_haendelse.rst
+   database_changes.rst
```
Comment (danchr) [inline doc/index.rst]: 🍰 👍

File: doc/database_changes.rst (hunk 6 above)
Comment (agger-magenta) [reply doc/database_changes.rst]: I've made a major revision of this document based on your comments. I believe it has improved on many points.

File: oio_rest/tests/test_integration_class.py (hunk 20 above)
Comment (agger-magenta) [reply oio_rest/tests/test_integration_class.py]: I added an integration test with PUT, in the Class integration suite. Christoffer asked for that as well.
